hoomd/md/ActiveForceConstraintCompute.h

```diff
@@ -0,0 +1,221 @@
+ orientation of any torque vector
+ * relative to the force vector is preserved
+    \param timestep Current timestep
+*/
+template<class Manifold>
+void ActiveForceConstraintCompute<Manifold>::rotationalDiffusion(uint64_t timestep)
+    {
+    //  array handles
+    ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(), access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::readwrite);
+    ArrayHandle<unsigned int> h_tag(m_pdata->getTags(), access_location::host, access_mode::read);
+
+    assert(h_pos.data != NULL);
+    assert(h_orientation.data != NULL);
+    assert(h_tag.data != NULL);
+
+    for (unsigned int i = 0; i < m_group->getNumMembers(); i++)
+        {
+        unsigned int idx = m_group->getMemberIndex(i);
+        unsigned int ptag = h_tag.data[idx];
+        hoomd::RandomGenerator rng(
+            hoomd::Seed(hoomd::RNGIdentifier::ActiveForceCompute, timestep, m_sysdef->getSeed()),
+            hoomd::Counter(ptag));
+
+        quat<Scalar> quati(h_orientation.data[idx]);
+
+        Scalar3 current_pos = make_scalar3(h_pos.data[idx].x, h_pos.data[idx].y, h_pos.data[idx].z);
+        Scalar3 norm_scalar3 = m_manifold.derivative(current_pos);
+        Scalar norm_normal = fast::rsqrt(dot(norm_scalar3, norm_scalar3));
+        norm_scalar3 *= norm_normal;
+        vec3<Scalar> norm = vec3<Scalar>(norm_scalar3);
+
+        Scalar delta_theta = hoomd::NormalDistribution<Scalar>(m_rotationConst)(rng);
+        Scalar theta
+            = delta_theta
+              / 2.0; // half angle to calculate the quaternion which represents the rotation
+        quat<Scalar> rot_quat(slow::cos(theta),
+                              slow::sin(theta) * norm); // rotational diffusion quaternion
+
+        quati = rot_quat * quati; // rotational diffusion quaternion applied to orientation
+        h_orientation.data[idx] = quat_to_scalar4(quati);
+        }
+    }
+
+/*! This function sets a manifold constraint for all active particles. Torque is not considered here
+ */
+template<class Manifold> void ActiveForceConstraintCompute<Manifold>::setConstraint()
+    {
+    //  array handles
+    ArrayHandle<Scalar4> h_f_actVec(m_f_activeVec, access_location::host, access_mode::read);
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::readwrite);
+    ArrayHandle<Scalar4> h_pos(m_pdata->getPositions(), access_location::host, access_mode::read);
+
+    assert(h_f_actVec.data != NULL);
+    assert(h_pos.data != NULL);
+    assert(h_orientation.data != NULL);
+
+    if (!m_manifold.fitsInsideBox(m_pdata->getGlobalBox()))
+        {
+        throw std::runtime_error("Parts of the manifold are outside the box");
```

Comment:
Most errors in HOOMD are done with the m_exec_conf objects. I believe this is to properly handle a variety of run environments that the std library may not.

Reply:
The use of `std::runtime_error` is correct here.

Exceptions are handled properly even in MPI and provide a cleaner message to the user. There is another big cleanup project in the future to convert `error << msg; throw runtime_error` to `throw runtime_error(msg)` in all of hoomd.